src/java/org/apache/cassandra/service/TruncateResponseHandler.java

```diff
@@ -61,24 +63,34 @@ public TruncateResponseHandler(int responseCount)
     public void get() throws TimeoutException
     {
         long timeoutNanos = getTruncateRpcTimeout(NANOSECONDS) - (nanoTime() - start);
-        boolean completedInTime;
+        boolean signaled;
         try
         {
-            completedInTime = condition.await(timeoutNanos, NANOSECONDS); // TODO truncate needs a much longer timeout
+            signaled = condition.await(timeoutNanos, NANOSECONDS); // TODO truncate needs a much longer timeout
         }
         catch (InterruptedException e)
         {
             throw new UncheckedInterruptedException(e);
         }
 
-        if (!completedInTime)
-        {
+        if (!signaled)
             throw new TimeoutException("Truncate timed out - received only " + responses.get() + " responses");
-        }
 
-        if (truncateFailingReplica != null)
+        if (!failureReasonByEndpoint.isEmpty())
         {
-            throw new TruncateException("Truncate failed on replica " + truncateFailingReplica);
+            // clone to make sure no race condition happens
+            Map<InetAddressAndPort, RequestFailureReason> failureReasonByEndpoint = new HashMap<>(this.failureReasonByEndpoint);
+            int size = failureReasonByEndpoint.size();
+            long timeouts = failureReasonByEndpoint.values().stream().filter(RequestFailureReason.TIMEOUT::equals).count();
+            long nonTimeout = size - timeouts;
+            if (nonTimeout <= timeouts)
+                throw new TimeoutException("Truncate timed out - received only " + responses.get() + " responses");
```

Comment:
What if instead of just throwing a `TimeoutException` when there are more timeouts than non-timeouts (okay >=), we just get rid of what's between lines 83 and 88 here? (i.e. We just throw a `TruncateException` with the per-endpoint failure summary snapshot, unless perhaps every single one of the failure reasons is TIMEOUT?)

Reply:
the bug reported was when 100% of the replies are timeout and we return the wrong exception type, so I wouldn't want to return a `TrucateException` in this case.

Now, if `signal=true && failed`, I guess we have a more fundmental question of "how many replicas timeout for us to call it a timeout"?  I would prefer we are consistent so would want all read/write/truncate to match the logic...

So, lets say we are dealing with a LOCAL_QUORUM read with rf=6, if failures=3 then we will fail, but lets say we have the following

R1: TIMEOUT
R2: TIMEOUT
R3: UNKNOWN

*should* this be a timeout, or should this be an error?  *should* this only ever be a timeout when 100% are timed out?  Personally I feel that if majority are timeout, we actually timed out.  @clohfink thoughts?

Reply:
If you pointed a gun to my head rn, for all three cases (read/write/truncate), I would just...
1.) throw a timeout if all the failure reasons in the map are TIMEOUT
2.) show the user the failure map by endpoint attached to a real error otherwise

Reply:
spoke with @clohfink and Caleb in slack and will go with 100% replies are timeout = timeout, this would then just make it so this race condition properly returns timeout and doesn't impact any other case.